nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java

```diff
@@ -0,0 +1,436 @@
+            "    if (checkAndReturn(o1 != null) && o2 != null) {",
+            "      return o1.toString() + o2.toString();",
+            "    } else {",
+            "      return o1.toString() + ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "        o2.toString();",
+            "    }",
+            "  }",
+            "  boolean test2(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 != null) && o1.toString().isEmpty();",
+            "  }",
+            "  boolean test3(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 == null) ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      && o1.toString().isEmpty();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePath() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "class Test {",
+            "  String test(Object required) {",
+            "    return Validation.identity(required == null)",
+            "      ? required.toString()",
+            "      : required.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePathAfterFailure() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o) {",
+            "    Validation.checkTrue(o == null);",
+            "    return Validation.identity(o == null)",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      ? o.toString()",
+            "      : o.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void complexContractUnreachablePathAfterFailure() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "class Test {",
+            "  @Contract(\"false, !null -> fail\")",
+            "  static void checkTrue(boolean value, @Nullable Object other) {",
+            "    if (!value) throw new RuntimeException();",
+            "  }",
+            "  String test(@Nullable Object o1, @Nullable Object o2, Object required) {",
+            // This case cannot be handled by the preferred conditional throw node insertion
+            // path due to reliance on nullness information from a second argument.
+            "    checkTrue(o1 == null, required);",
+            // o1 is guaranteed to be null based the previous check, allowing
+            // us to understand this expression as 'null == o2'.
+            "    return Validation.identity(o1 == o2)",
```

Comment:
Ok, in order for this to work, we need explicit tracking of `o1` as `Nullness.NULL` not `Nullness.NULLABLE`, and we seem to have it here, but I am not sure how deep this goes... a bit worried about how well we handle this in general. What if these are more complex APs than simple variables? (map gets, getter calls, etc)